url1a41b8e9f2_mr1azl_INF345_tgz-pJ8-TestParserJ8/src/xx/XPathParser.java

```diff
@@ -0,0 +1,3269 @@
+        }
+        else {
+            {if (true) return compiler.qname(nc1, nc2);}
+        }
+    throw new Error("Missing return statement in function");
+  }
+
+  final public Object parseExpression() throws ParseException {
+    Object ex;
+    ex = Expression();
+    jj_consume_token(0);
+        {if (true) return ex;}
+    throw new Error("Missing return statement in function");
+  }
+
+/* [1] LocationPath ::= RelativeLocationPath | AbsoluteLocationPath  */
+  final public Object LocationPath() throws ParseException {
+ Object ex = null;
+    switch ((jj_ntk==-1)?jj_ntk():jj_ntk) {
+    case OR:
+    case AND:
+    case MOD:
+    case DIV:
+    case NODE:
+    case TEXT:
+    case COMMENT:
+    case PI:
+    case AXIS_SELF:
+    case AXIS_CHILD:
+    case AXIS_PARENT:
+    case AXIS_ANCESTOR:
+    case AXIS_ATTRIBUTE:
+    case AXIS_NAMESPACE:
+    case AXIS_PRECEDING:
+    case AXIS_FOLLOWING:
+    case AXIS_DESCENDANT:
+    case AXIS_ANCESTOR_OR_SELF:
+    case AXIS_FOLLOWING_SIBLING:
+    case AXIS_PRECEDING_SIBLING:
+    case AXIS_DESCENDANT_OR_SELF:
+    case FUNCTION_LAST:
+    case FUNCTION_POSITION:
+    case FUNCTION_COUNT:
+    case FUNCTION_ID:
+    case FUNCTION_KEY:
+    case FUNCTION_LOCAL_NAME:
+    case FUNCTION_NAMESPACE_URI:
+    case FUNCTION_NAME:
+    case FUNCTION_STRING:
+    case FUNCTION_CONCAT:
+    case FUNCTION_STARTS_WITH:
+    case FUNCTION_CONTAINS:
+    case FUNCTION_SUBSTRING_BEFORE:
+    case FUNCTION_SUBSTRING_AFTER:
+    case FUNCTION_SUBSTRING:
+    case FUNCTION_STRING_LENGTH:
+    case FUNCTION_NORMALIZE_SPACE:
+    case FUNCTION_TRANSLATE:
+    case FUNCTION_BOOLEAN:
+    case FUNCTION_NOT:
+    case FUNCTION_TRUE:
+    case FUNCTION_FALSE:
+    case FUNCTION_NULL:
+    case FUNCTION_LANG:
+    case FUNCTION_NUMBER:
+    case FUNCTION_SUM:
+    case FUNCTION_FLOOR:
+    case FUNCTION_CEILING:
+    case FUNCTION_ROUND:
+    case FUNCTION_FORMAT_NUMBER:
+    case NCName:
+    case 82:
+    case 83:
+    case 86:
+    case 88:
+      ex = RelativeLocationPath();
+      break;
+    case SLASH:
+    case SLASHSLASH:
+      ex = AbsoluteLocationPath();
+      break;
+    default:
+      jj_la1[5] = jj_gen;
+      jj_consume_token(-1);
+      throw new ParseException();
+    }
+        {if (true) return ex;}
+    throw new Error("Missing return statement in function");
+  }
+
+/* [2] AbsoluteLocationPath ::= '/' RelativeLocationPath? | AbbreviatedAbsoluteLocationPath  */
+/* [10]    AbbreviatedAbsoluteLocationPath    ::=    '//' RelativeLocationPath  */
+  final public Object AbsoluteLocationPath() throws ParseException {
+    ArrayList steps = new ArrayList();
+    if (jj_2_2(2147483647)) {
+      LocationStep(steps);
+      label_1:
+      while (true) {
+        switch ((jj_ntk==-1)?jj_ntk():jj_ntk) {
+        case SLASH:
+        case SLASHSLASH:
+          ;
+          break;
+        default:
+          jj_la1[6] = jj_gen;
+          break label_1;
+        }
+        LocationStep(steps);
+      }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 434:446, 466:478.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.